### PR TITLE
Setup e2e tests

### PR DESCRIPTION
**Pull request by tyitang**
## Description

1. Setup e2e tests with hardhat, typescript, and mocha
2. Add the test dependencies
3. Add sample tests for IP assets, groups, and permissions

## Notes

- Example: Links and navigation need to be added to the front-end interface

### CODE REVIEW

File: test/hardhat/e2e/constants.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+export const AccessController = "0xf709c8001E94e2ca6F98b7fFBCd5BD3943E46D81";
```
Comment (limengformal) [inline test/hardhat/e2e/constants.ts:1]: Add a comment at the file level to describe what these constants are

Comment (limengformal) [reply test/hardhat/e2e/constants.ts:1]: Also, are these addresses fixed for all deployment?

File: test/hardhat/e2e/grouping/group.ipa.test.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    expect(groupId).to.be.properHex(40);
+  });
+
+  it("Register Group IPA with non-whitelisted group pool", async function () {
+    const nonWhitelistedGroupPool = "0xC384B56fD62d6679Cd62A2fE0dA3fe4560f33300"
+    await expect(
+      this.groupingModule.registerGroup(nonWhitelistedGroupPool)
+    ).to.be.rejectedWith(Error)
+  });
+});
```
Comment (limengformal) [inline test/hardhat/e2e/grouping/group.ipa.test.ts:22]: Add empty line at the end

File: test/hardhat/e2e/grouping/group.ipa.test.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import "../setup"
```
Comment (limengformal) [inline test/hardhat/e2e/grouping/group.ipa.test.ts:1]: Add a file level comment to describe the file

File: test/hardhat/e2e/ipa/ipa.test.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import "../setup"
```
Comment (limengformal) [inline test/hardhat/e2e/ipa/ipa.test.ts:1]: Add a file level comment to describe the file

File: test/hardhat/test.js
Comment (limengformal) [inline test/hardhat/test.js:1]: @kingster-will can you see if this file is needed?

**Review by limengformal — COMMENTED** [review]
@tyitang For all new files, add a file level comment to describe what the file does.

**Comment by limengformal** [comment]
@tyitang Add a README.md file in test/hardhat folder to describe how to run the tests.

File: test/hardhat/e2e/constants.ts (hunk 1 above)
Comment (tyitang) [reply test/hardhat/e2e/constants.ts:3]: They are fixed except the MockERC721 address, I moved it to env file

File: test/hardhat/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+```sh
+yarn # this installs packages
+```
+
+You'll need to add the variables refer to the .env.example to a .env file at project root.
+
+Then, at project root run the tests with command:
+
+```sh
+npx hardhat test --network odyssey
```
Comment (limengformal) [inline test/hardhat/README.md:25]: Add the command to run test on each test file as well

**Review by limengformal — APPROVED** [review]

**Pull request merged**